lib/rouge/lexers/armasm.rb

```diff
@@ -0,0 +1,82 @@
+# -*- coding: utf-8 -*- #
+# frozen_string_literal: true
+
+module Rouge
+  module Lexers
+    class ArmAsm < RegexLexer
+      title "ArmAsm"
+      desc "Arm assembly syntax"
+      tag 'armasm'
+      filenames '*.s'
+
+      state :root do
+        rule /\n/, Text
+        rule /^[ \t]*#[ \t]*((define|elif|else|endif|error|if|ifdef|ifndef|include|line|pragma|undef|warning)[ \t].*)?\n/, Comment::Preproc
+        rule /[ \t]+/, Text, :command
+        rule /;.*\n/, Comment
+        rule /\$[A-Za-z_][0-9A-Za-z_]*\.?/, Name::Namespace, :afterlabel # variable substitution or macro argument
+        rule /([0-9A-Za-z_][0-9A-Za-z_]*|\|[^|\n]+\|)/, Name::Label, :afterlabel
+        end
+
+      state :afterlabel do
+        rule /\n/, Text, :root
```

Comment:
The states defined in a lexer operate as items in a stack. If you're leaving a state to return to the `:root` state, you should pass the special state, `:pop!`, which will cause the current state to be popped off the stack and for the lexer to return to the previous state.

If it doesn't make sense to leave the current state, you just want to incorporate the rules from another state, you can do this by calling `mixin :<state>`. This will cause the rules in <state> to be 'copied' in at the location of the `mixin` call.

Reply:
OK. A stack of states doesn't particularly suit this syntax, as the state is reset at the start of every line, and which states you can transition to depend largely on how much whitespace has preceded it on the same line. Is there any way of implementing a state variable that doesn't stack its previous value, or alternatively a way of popping all states off the stack when I reach the end of a line?

Reply:
You can still use a state stack in that instance. It's just that whenever you match something that ends the line, you'd pass the state `:pop!`.

You can also move between states by using `goto` rather than `push`. `goto` automatically pops the current state off the stack and swaps it for the new state.

Reply:
Perhaps another way to think about it is as construct. A simple string might not span multiple lines but we might still lex it by using a separate state. Here's a naïve implementation:

```rb
state :string do
  rule /[^"]+/, Str
  rule /["]/, Str, :pop!
end
```

Hope that helps make things clearer!